File: python/reader_util.py
def bracket_reader(string: str, type_of_brace: str) -> list:
    """get to know where a bracket starts and is successfully closed"""
    braces = {'{': '}', '[': ']', '(': ')'}

    if type_of_brace not in braces:
        raise TypeError(f'{type_of_brace} is not supported')

    brace = type_of_brace
    partner_brace = braces.get(type_of_brace)
    brace_positions = []
    position = 0

    # get the position of the first brace
    try:
        start = string.index(brace, position)
    except ValueError:
        start = None

    while start:
        end_position = None
        opening_brace_count = 1

        # loop through the whole string and find the closing brace
        for index, strand in enumerate(string):
            # brace count can only be changed if the loop is beyond index start
            if index > start:
                # if it's another opening brace increment opening_brace_count
                if strand == brace:
                    opening_brace_count += 1

                # if it's a closing brace decrement opening_brace_count
                elif strand == partner_brace:
                    opening_brace_count -= 1

            # when opening_brace_count is 0, it means that an opened bracket is
            # now closed and there's no need to proceed with the loop
            # create an new position to find a new opening brace
            if opening_brace_count == 0:
                end_position = index
                position = index
                break

        # if it does not close raise a value error
        if opening_brace_count:
            raise ValueError('Brackets do not match')

        # add the brace to brace positions [start, end_position] => [[start, end_position]]
        brace_positions.append([start, end_position])

        # check if there's a new brace?
        try:
            start = string.index(brace, position)
        except ValueError:
            start = None

    # return a list of substrings of string with an open and corresponding closing brace
    return [
        string[
            brace_position[0]:brace_position[-1] + 1
        ].translate(str.maketrans('\n', ' ')) for brace_position in brace_positions
    ]

File: ts/readerUtil.ts
export const brackets = {
    ROUND_BRACKET: '(',
    SQUARE_BRACKET: '[',
    CURLY_BRACKET: '{',
    ANGLE_BRACKET: '<'
};


const patnerBraces = new Map(
    [
        [brackets.ROUND_BRACKET, ')'],
        [brackets.SQUARE_BRACKET, ']'],
        [brackets.CURLY_BRACKET, '}'],
        [brackets.ANGLE_BRACKET, '>']
    ]
);


export function bracketReader(stringToRead: string, braceToRead: string): string[] {
    /* 
    get to know where a bracket starts and is successfully closed
    */

    const isBraceAllowed = (_brace: string): _brace is keyof typeof brackets => patnerBraces.has(_brace);

    if (!isBraceAllowed(braceToRead)) {
        throw TypeError(
            `'${braceToRead}' is not supported. Import the constant brackets and use 'brackets.${Object.keys(brackets).join('\', \'brackets.')}'`
        );
    };

    const partnerBrace = patnerBraces.get(braceToRead);
    let bracePositions: number[][] = [];
    const stringToReadAsList = [...stringToRead];
    let position = 0;

    const getItemIndex = function (item: string, index: number): Boolean {
        let result = false;
        if (index >= position) {
            if (item === braceToRead) {
                result = true;
            }
        }
        return result;
    };

    let start = stringToReadAsList.findIndex((item, index) => getItemIndex(item, index));

    while (start > -1) {
        let endPosition: number = Infinity;
        let openingBraceCount = 1;


        // loop through the whole string and find the closing brace
        for (let index in stringToReadAsList) {
            const intIndex = parseInt(index);
            let strand = stringToReadAsList[intIndex];

            // a brace can only be closed if we are beyond index 'start'
            if (intIndex > start) {
                if (strand === braceToRead) {
                    // if it's anothe opening brace increment openingBraceCount
                    openingBraceCount++;
                } else if (strand === partnerBrace) {
                    // if its a closing brace then decrement openingBraceCount
                    openingBraceCount--;
                };
            };
            // when opening_brace_count is 0, it means that an opened bracket is
            // now closed and there's no need to proceed with the loop
            // create an new position to find a new opening brace
            if (openingBraceCount == 0) {
                endPosition = intIndex;
                position = intIndex;
                break;
            };
        };

        // if it does not close raise a value error
        if (openingBraceCount) {
            throw new Error(`No closing brace found!`);
        };

        // add the brace to brace positions [start, end_position] => [[start, end_position]]
        bracePositions.push([start, endPosition]);

        // check if there's a new brace?
        start = stringToReadAsList.findIndex((item, index) => getItemIndex(item, index));
    };

    // return a list of substrings of string with an open and corresponding closing brace
    return bracePositions.map(
        bracePos => stringToRead.substring(bracePos[0], bracePos[1] + 1).replace('\n', '')
    );
};